### ISSUE TITLE
partition and unpartition

### ISSUE THREAD

**Issue by QirongZhu** [exafmm/exafmm#34]
Hi, I am wondering whether it is easy to add a feature to revert the partition operation such that bodies are sent back to where they came from after FMM step is completed. I noticed that exafmm-alpha has such a feature(?). I am calling exafmm for a selection of N-body particles and the number of active particles varies from one timestep to the next timestep. Therefore, it seems necessary to receive unpartitioned bodies to get forces/potentials to the same MPI rank.

**Comment by rioyokota** [comment]
Yes, this is very easy to do. You just need to save the rank where each body came from and send it back to that rank using alltoallv. This is what exafmm-alpha does.

**Comment by QirongZhu** [comment]
Thanks for the prompt reply. It'd be great if the new pushes can include this feature.

**Comment by rioyokota** [comment]
I have been intermittently working on the new partitioning and local essential tree communication scheme in the new exafmm, but haven't had the time to fully debug it. Since I have very little time to work on this, it might be faster if you could contribute to the parallel scheme of exafmm.

Here is how to run the new exafmm:
git checkout dev
./configure
cd 3dp_simd_mpi
make fmm
mpirun -np 2 ./fmm -d p

You will see a list of numbers being printed on the screen
This is showing the 
{MPIRANK} {Index of target cell} {Level of target cell} {Index of source cell} {Level of source cell}
when the bodies are missing from the source cell even when it is required by the horizontalPass

This happens because the local essential tree communication is not sending what should be sent.
The source code for the local essential tree is in local_essential_tree.h.
Could you please take a look at it with a fresh pair of eyes.
You might be able to find the bug.

By the way, I have switched from the orthogonal recursive bisection scheme in exafmm-alpha to the hashed octree approach in exafmm, because the ORB caused the interaction list to grow. I think the hashed octree is a much more scalable approach. However, it is very difficult to determine what to send for the local essential tree. I take a similar approach to the Bonsai code, so you can take a look at their paper for the details of how the hashed octree partitioning and local essential tree communication is done.
[2014Bedorf,J.pdf](https://github.com/exafmm/exafmm/files/1529463/2014Bedorf.J.pdf)

**Comment by QirongZhu** [comment]
Sure, I will try to see what is going on there. Thanks for the invitation and the references.

**Comment by QirongZhu** [comment]
Hi, is selectCells() function used to skip sending particle data if this cell is well separated from another task? 

It appears in the function getDistance(cell, rank), it mostly returns 0, therefore we are sending bodies data anyway. Will there be over-lapping/intersecting regions between two tasks? If not, the distance calculation might be off. One possible bug is that 'norm' function in getDistance(cell c, rank) needs be followed by a square root since we are using 
"R2 = R * R * theta * theta". 

It also seems tree walk in the horizontal pass are not the same from selectCells(). I compared the radius of source and sink cells. The error message is likely caused by two cells with very different sizes. Presumably, if a very small cell can pass 
"R2 > (Cj->R + Cj->R) * (Cj->R + Cj->R)", 
it still cannot pass 
"R2 > (Cj->R + Ci>R) * (Cj->R + Ci->R)" if Ci->R >> Cj-R. 
As a result, bodies data is missing.

**Comment by rioyokota** [comment]
getDistance is for calculating the distance between a source cell in your rank and the edge of the bounding box of the entire partition of a remote rank. Therefore if you are examining your own rank the distance should be zero. I should change this so that it skips your own rank and only examines remote ranks.

You are correct that R2 > (Cj->R + Cj->R) * (Cj->R + Cj->R) will not catch cases where Ci->R >> Cj->R, and this seems to be where it is failing. There is a sublte difference in how R2 is being calculated in the two cases though. Since we can't see the size and positon of the target cell, we need to account for the worst case. That would be to compare with the largest possible cell in that remote domain. I think we should put a restriction on the difference of sizes that interacting cells can have. For example restrict it to at most one level apart. This means large leafs need to be forcefully divided even if the number of bodies is smaller than ncrit. Otherwise, there will always be a large leaf that tries to interact with much smaller cells.

So far people have resorted to the strategy of simply sending more cells and bodies than necessary, just to be safe. This is obviously suboptimal since the amount of communication is always the bottleneck at the end of parallel scalability. To come up with a more scalable FMM, we need to put tighter bound on the local essential tree communication. I think this can be achieved by restricting the differences of sizes of interacting cells, and then using the worst case (one level larger Ci->R with the R2 being calculated for the case where Ci is right on the boundary of that remote partition).

It would be great if you can look into this problem.

**Comment by QirongZhu** [comment]
Sure, I will take a look at this and test the idea of using the strategy of restricting cell size differences you have outlined above.

**Comment by rioyokota** [comment]
Hi Qirong, I would like to add you to our exafmm slack channel. Do you use Slack?

**Comment by QirongZhu** [comment]
Hi, Thanks a lot for the invitation. Please use my Gmail address for Slack: zhuqirong1874@gmail.com. I am not familiar with Slack, though.
